lib/spack/spack/util/path.py

```diff
@@ -168,6 +169,42 @@ def sanitize_file_path(pth):
     return drive + os.path.join(*pth)
 
 
+def args_to_pathlib(*args):
+    return [
+        pathlib.Path(arg) if (isinstance(arg, str) and not is_path_url(arg)) else arg
+        for arg in args
+    ]
+
+
+def pathlib_filter(_func=None, arg_slice=None):
```

Comment:
That would certainly be useful, without a doubt in the case where we want to model that an abstract operation involving paths is being performed without direct interaction with the filesystem. I was under the impression however, that in general the filesystem and related modules are for interacting directly with the filesystem, or if not directly, transitively from the calling context, so I'm not sure how much use that would have. IMO even if the method invoking the decorator may not have direct interaction with the filesystem, if its returned value is intended to model a path on the filesystem (interacting or not) it should be a `Path` not a `PurePath`.
All that said, some nuance here is 100% a great idea. Will implement in this PR (if not here then where right?), although I may not apply it very liberally.

Reply:
Please feel free to *not* implement that in this PR if it ends up increasing scope too much! This could definitely be left as a TODO as well. Thanks so much for this thoughtful response!

Reply:
Already implemented, it was fairly trivial. Will be included in next push. Keep getting pulled away before I can clean everything up.

Reply:
No rush!!